
############################################################################################################
#
#   Objective: Integrated TopMed WGS mCA calling pipeline from raw vcf mCAta to mCA instances using MoChA.
#
#              This R script is required. It was designed to merely keep the mCA phenotypes when specific 
#              conditions are satisfied. 
#              
#   NOTE: Please refer to https://github.com/freeseek/mocha for GRCh38 resources.
#         In this pipeline, we assume they are ready to be applied.
#
#   	  This pipeline is contingent to be tweaked according to your research requirements.
#
#         All required/upmCAted softwares should be installed before running this pipeline; otherwise,
#         it will cause errors. 
#
#         bcftools(1.11 or later), shapeit4(4.2.2 or later, if VCF needs phasing), bcftools-mocha(1.15 or later), 
#         bedtools(2.30 or later)
#          
#
############################################################################################################
#   
#   Revision History: 
#
#   mCAte              Author               Title                       Modification
#   09-23-2023        Mr.Xiaolong Ma       Bioinformatics Analyst      v1
#   
#   Contact			  Institute
#   xima@mcw.edu      Medical College of Wisconsin
#
############################################################################################################
#
#   Directory Hierarchy
#
#   your_proj_working_directory
#   .
#   ├── GRCh38              ----> to save the prepared GRCh38 resources 
#   ├── scripts 			----> to save the bash scripts  
#   ├── mCA             	----> to save the mChA calling result files
#   |── raw_mCAta            ----> to store the input vcf/bcf raw mCAta 
#	|── logs				----> to save logs for checkup
#   ├── mis    				----> to save sample list, metamCAta and miscellaneous		
#
#
#   Input: A single .tsv composed of all mCA phenotypes, which could be generated by optional_step2_mCA_compile.sh
#
#
############################################################################################################
tsv <- "path_to_this_tsv/all.mCA.tsv"

mCA <-read.csv(file=tsv, header=T,sep="\t")

# mCA to be excluded
# 1) has chrX but gender is not well infered
# 2) n_flips = -1         
#    n_flips - number of phase flips for calls based on BAF and genotype phase model (-1 if LRR and BAF model used) 

f1 <- (mCA$chrom == "chrX" & (mCA$computed_gender=="U" | mCA$computed_gender=="M")) | mCA$n_flips == -1
mCA <-mCA[!f1, ]


# select long mCA
f2<- mCA$n_hets > 2000
f3<- mCA$rel_cov > 2.9
f4<- mCA$rel_cov > 2.5 & mCA$bdev > 0.16
f5<- mCA$lod_baf_phase > 5
f6<- mCA$bdev > 0

f_all <- f2 & !f3 & !f4 & f5 & f6

output<-"path_to_output/long.mCA.filtered.tsv.gz"

write.table(mCA[f_all,], gzfile(output), col.names= T, row.names=F,sep="\t", quote=F)

